documentation/dataplane/dash-bmv2-data-plane-app.md

```diff
@@ -0,0 +1,349 @@
+  |                       Action                                  |
+  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+  |                       Meter Class                             |
+  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+                                                                 
+  |                       [overlay rewrite]                       :
+  :                                                               |
+  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+  |                       [tunnel encap]                          :
+  :                                                               |
+  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+  |                       [underlay encap]                        :
+  :                                                               |
+  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+                                                                   
+  D: flow direction, 0, oubound; 1, inbound                                             
+  Version: flow version                               
+  Action: flow data – action
+  Meter: Meter class
+  ...
+```
+
+### 6.2. Basic Flow
+
+Basic flow means basic operation of flow, i.e. CRUD. Generally DPAPP may run a flow state machine to manage flow state and do proper operation of CRUD. The state transformation is triggered by these events:
+- Packets with DASH metadata from pipeline,  for flow creation/deletion/refresh
+- Timer, for flow age-out
+
+The below sequence chart shows the detail steps of flow creation, flow state becomes FLOW_CREATED. Other operations are similar.
+
+```mermaid
+sequenceDiagram
+    participant P as P4 Pipeline
+    participant R as P4 Runtime
+    participant W as VPP worker
+    participant M as VPP master
+    autonumber
+    P->>+W: DASH metadata (source PIPELINE) + customer packet
+    W->>+W: Create flow in local flow table
+    W->>+M: Request current flow offload
+    M->>+R: Call DASH SAI create_flow (gRPC)
+    R->>R: Write flow to P4 flow table
+    R->>M: create_flow returns OK
+    M->>W: Reply flow offload done
+    W->>P: DASH metadata (source DPAPP) + customer packet
+```
+
+It is remarkable that VPP worker should not call DASH SAI API directly, otherwise DASH SAI may block VPP worker to handle other packets. Each flow has a packet queue. In step 3, it enqueues current packet firstly and then requests flow offload. In step 8, it dequeues the packet and then sends the packet back to P4 pipeline.
+
+### 6.3. Flow resimulation
+
+```mermaid
+sequenceDiagram
+    participant C as DASH SAI Client
+    participant P as P4 Pipeline
+    participant R as P4 Runtime
+    participant W as VPP worker
+    participant S as DASH SAI Server
+    autonumber
+    C->>+S: ENI_ATTR_FULL_FLOW_RESIMULATION_REQUESTED
+    S->>+R: Call DASH SAI set_eni_attr (epoch) 
```

Comment:
I believe we don't have set eni attr to update epoch. we will need to call p4 runtime api directly or something to make it work.

Reply:
also I believe epoch is named as flow version.

Reply:
Add epoch as an internal attribute of eni and flow, which is invisible to public SAI.